### DISCUSSION TITLE
Repeater device

### DISCUSSION THREAD

**Discussion by ghoti57** [ghoti57/evofw3#27] [Ideas]
Before beginning a discussion on a repeater device I'll define some terminology to avoid confusion.

**Repeater**: A device that re-transmits packets it has received
**Originator**: A device that creates and transmits the original copy of a packet
**Listener**: A device for which an Originator packet is intended.
**Proxy**: A device that filters received packets, modifies the addresses in those packets and then transmits the modified packet.

**CCA**: Clear Channel Assessment - an indication that no devices are currently transmitting

**Repeater** devices require no configuration.  They retransmit identical copies of every packet they receive.
**Proxy** devices require configuration of the address translations required.

This discussion is specifically about a **Repeater** device.

I'll edit this entry with new terms as necessary

**Comment by ghoti57** [comment]
The biggest problem with implementing a repeater on any device is recognising when it is safe to retransmit a packet.

This requires a CCA indication.

Strictly speaking this is required in the standard Gateway application that is the "HGI80" emulation application currently implemented but since the origination of messages in that environment is more asynchronous it seems to be less critical.

The problem in a Repeater is that some Originator devices transmit multiple copies of messages.
These copies are synchronised with each other and collisions between the Originator copies and Repeater re-transmission is significantly more likely causing failure of both.

**Comment by ghoti57** [comment]
Once CCA indications are available implementation of a single Repeater device is fairly trivial within evofw3.

Multiple Repeaters in a single system are a little harder as you must avoid ping-ponging messages between Repeaters indefinitely.

**Reply by pcr20** [reply]
Would a suitable CCA indicator simply by a random time back-off algorithm? (I know this approach is common in some ISM band applications)
eg "1. CCA set to 'false' 2. receive message 3. wait random amount of time 4. CCA set to 'true' 5. retransmit"

There could be additional logic to not attempt to retransmit multiple copies of the same message (eg introduce a sent queue where there can only be one copy of a specific message in the queue)

Any idea how this device [BRDG-02M11](https://ithodaalderop.compano.com/Data/Environments/000001/Attachment/Bijlage/A04_Ventilatie/A04_80_Accessoires/A04_80_99_Overig/A04_80_99_01_Overig/B01_06_MTE/Manual%20repeater%20BRDG-02M11.pdf) works? Is it truely a Repeater or is it more like a Proxy?

**Reply by ghoti57** [reply]
The problem with random back-off is the delay cannot be for very long.  Some of the Listener devices are battery operated and they are only awake for short periods of time.  It's important that the retransmitted messages go out in a timely fashion.

**Reply by ghoti57** [reply]
BRDG-02M11 is a Repeater.  It is selective in which messages it retransmits because it will only forward packets between devices if it has seen them bind.  Neither of the bound devices are aware that the BRDG is present.  If it were a proxy it would have to participate in the binding process.

You would almost certainly perform the binding with all three devices in close proximity before moving one of them to its remote location afterwards.

**Reply by ghoti57** [reply]
On the CCA back-off,  evofw3 has limited buffer space.

You need to retransmit each packet as soon as possible.
CCA timers would have to restart if you received a new packet while being backed off.

**Reply by pcr20** [reply]
Is this maths roughly right?:

messages: ~100 bytes (ie ~1000 bits)
baud rate: ~38400bit/s ??
message length: 1/38s = 26ms

HR92s transmit ~every 8 minutes = every 480s (read this in a forum post I seem to recall?)
how long do they leave the rx window open for after transmit? 100ms?

**Reply by pcr20** [reply]
> BRDG-02M11 is a Repeater. It is selective in which messages it retransmits because it will only forward packets between devices if it has seen them bind. Neither of the bound devices are aware that the BRDG is present. If it were a proxy it would have to participate in the binding process.
> 
> You would almost certainly perform the binding with all three devices in close proximity before moving one of them to its remote location afterwards.

BDRG architecture sounds a reasonable architecture to copy then? any idea how its timing works?

**Reply by ghoti57** [reply]
> 
See wiki for s discussion for a discussion about packet length and transmission times.

**Comment by pcr20** [comment]
@ghoti57 is there not also an issue that the Originator may behave unexpectedly  if it received a packet sent by a repeater? (it is presumably not designed to receive a packet which appears to be one it has sent?)

You comment above that "some Originators send multiple copies of messages", therefore, is it the case that the Rameses protocol allows a Listener to receive multiple copies of the same message?

**Reply by ghoti57** [reply]
I don't think originators will have any problem discarding copies of its own messages, they will not be addressed in a way that will cause issues.

The repeated messages always contain the same information so there's no conflict if a Listener receives multiple copies.

**Comment by ghoti57** [comment]
HR92s listen approximately every 3-4 minutes for controller sync messages (that tell them when to wake up next)  They then listen for the zone control messages that they need.  In systems with many zones these messages actually span multiple packets which aggravates the standoff timing approach to CCA.  

I suspect the HR92 stops listening as soon as it has received all its required data.  I also suspect it keeps listening if it misses anything.

Trying to guess anything else about its behaviour is risky.

**Comment by ghoti57** [comment]
I think a concerted effort to get the CC1101 CCA indication working is the correct thing to try first.

All the documentation says it works.  I'm sure that's how the evohome devices will work.

**Reply by pcr20** [reply]
Have  you tried the CC1101 CCA already? (doesn't look super hard, but I can see it being a pain to get working without a scope/receiver):

17.5 Clear Channel Assessment (CCA)
The Clear Channel Assessment (CCA) is used to indicate if the current channel is free or busy. The current CCA state is viewable on
any of the GDO pins by setting IOCFGx.GDOx_CFG=0x09. MCSM1.CCA_MODE selects the mode to use when determining CCA.
When the STX or SFSTXON command strobe is given while CC1101 is in the RX state, the TX or FSTXON state is only entered if the clear channel requirements are fulfilled. Otherwise, the chip will remain in RX. If the channel then becomes available, the radio will not enter TX or FSTXON state before a new strobe command is sent on the SPI interface. This feature is called TX-if-CCA. Four CCA requirements can be programmed:
- Always (CCA disabled, always goes to TX)
- If RSSI is below threshold
- Unless currently receiving a packet
- Both the above (RSSI below threshold and not currently receiving a packet)

**Comment by ghoti57** [comment]
I've managed to get the cc1101 to generate a signal when another device is transmitting.  Now I just have to work out how to use it :-)

**Reply by pcr20** [reply]
I'm happy to try it out - is there a branch you're working on? Also... sorry dumb Arduino newbie question... is there a schematic for your SSM-D board, or some instructions showing where the ICSP connections are so I can program/debug it (or is it possible to debug over the USB port)?

**Comment by ghoti57** [comment]
You program the SSM-D from the Arduino GUI over the USB link.

 

Debug is very difficult as you can’t litter the code with “printf”.
I haven’t tried to debug over an ISP connection, again I suspect it would be too destructive to the timing..

The only way I’ve found to debug is with a logic analyser monitoring the cc1101 SPI/GDO0/GDO2 signals together with using spare digital outputs to signal significant events from the firmware.  

All the pins available from the atmega32u4 module are on the header connectors along the edge of the carrier PCB and are labelled there.  You’ll have to cut the sleeve off if you want to use them.  Keep the sharp edge of any blade turned away from the boards to avoid damaging tracks.



The debug code that’s still in the source needs tweaking before it can be used on the SSM-D as the outputs it’s trying to use aren’t available on the PCB headers.

To see the Carrier Sense signal from the firmware it needs to be generated on GDO0 from the cc1101.  This pin is currently used to control loading the TX fifo for TX mode.  

 

The problem is how to use GDO0 for both functions.  That requires a rework of the code that switches to TX.  It needs to do it as quickly as possible after it has determined it is safe to do so (not necessarily as soon as Carrier Sense clears).



As quickly as possible probably means that it’s necessary to prime the TX fifo with data before verifying it’s OK to transmit.  That’s where the conflict in GDO0  lies.


GDO2 is used to present the RX data signal to the RX port of a UART in the atmega device.

**Reply by pcr20** [reply]
That makes sense.

Could you share the code snippet to enable the CCA line? (I assume some reg writes over SPI to the CC1101 - would save me trawling through the data sheet).

I'm considering trying to port evofw3 to a ESP8266/32 which would mean they would be plenty of ram/CPU to handle fast state switching. The ESP appear to have fairly comprehensive hw SPI support - i think with DMA but I've not read up fully yet.

It might provide an easier route to a repeater. As an aside BTW I did buy a Honeywell Ramses repeater from Itho Daalderop, which appeared to work for about 12 hours and now doesn't work.. it's clearly stuck in some warning state which is undocumented.

**Comment by ghoti57** [comment]
The initial changes you want are in the CarrierSense branch.

I've almost got repeater logic working but I've not quite got something right when there are multiple received packets to be retransmitted (Repeater specific changes not in github yet)

**Comment by ghoti57** [comment]
I've just got one of these to play with 
<https://thepihut.com/products/m5stack-m5stamp-c3-mate-with-pin-headers>
